### PR TITLE
add year and phone number for user details (DD review)

### PR DESCRIPTION
**Pull request by raniarusly**

### CODE REVIEW

File: src/main/webapp/form.html [hunk 1]
```diff
@@ -20,8 +20,17 @@ <h3>a recommendation system for students, by students!</h3>
     <form id="user-form" method="POST" enctype="multipart/form-data">
       <label for="Name">Hello! my name is</label>
       <input type="text" name="name" class="name" required></input>
-      <label for="Department">I am currently a student in the department of</label>
+      <label for="Department">I am a student in the department of</label>
       <input type="text" name="department" class="department" required></input>
+      <label for="year">currently on year</label>
```
Comment (lornaarmstrong) [inline src/main/webapp/form.html]: nit: rephrase this to "currently in year"

File: src/main/webapp/my-profile.html [hunk 2]
```diff
@@ -55,13 +55,22 @@ <h2>Edit Profile</h2>
               <input type="text" id="name" name="name" size="40" required></input>
               <label for="Department">I am currently a student in the department of</label>
               <input type="text" id="department" name="department" size="40" required></input>
+              <label for="year">currently on year</label>
```
Comment (lornaarmstrong) [inline src/main/webapp/my-profile.html]: nit: rephrase to "currently in year"

File: src/main/webapp/scripts/my-profile.js [hunk 3]
```diff
@@ -39,15 +39,18 @@ function loadDetails() {
...
+      const phoneHTML = "<h5>" + user.phone + "</h5>";
       const bioHTML = "<p>" + user.bio + "</p>";
-      document.getElementById("details").innerHTML = profilePictureHTML + nameHTML + departmentHTML + bioHTML;
+      document.getElementById("details").innerHTML = profilePictureHTML + nameHTML + departmentHTML + phoneHTML + bioHTML;
 
       // Pre-fill input for editting profile with the current user details
       document.getElementById("name").setAttribute("value", user.name); 
-      document.getElementById("department").setAttribute("value", user.department); 
+      document.getElementById("department").setAttribute("value", user.department);
+      document.getElementById("year").getElementsByTagName('option')[(user.year - 1)].selected = "selected";
```
Comment (lornaarmstrong) [inline src/main/webapp/scripts/my-profile.js:52]: quick question (just to help my understanding) - why do we subtract 1 here?

File: src/main/webapp/styles/common.css [hunk 4]
```diff
@@ -26,7 +26,21 @@ h3 {
...
   font-family: 'Questrial', sans-serif;
   font-size: 28px;
-  margin: 5px;
+  margin: 10px;
+}
+
+h4 {
+  color: white;
+  font-family: 'Questrial', sans-serif;
+  font-size: 22px;
```
Comment (lornaarmstrong) [inline src/main/webapp/styles/common.css]: nit: it seems a little odd to have two different headers with the same formatting (e.g. same font size). If they need to be different headers, then they should probably be different in format.

**Review by lornaarmstrong — COMMENTED** [review]
A few minor comments but looks great overall! Could you also include a screenshot showing any UI changes?

File: src/main/webapp/scripts/my-profile.js (hunk 3 above)
Comment (raniarusly) [reply src/main/webapp/scripts/my-profile.js:52]: the options are 0-indexed so 
year 1 is the first option, thus index 0
year 2 is the second option, thus index 1, etc
hope this explains it :)

Comment (lornaarmstrong) [reply src/main/webapp/scripts/my-profile.js:52]: ah I see, thanks!

**Review by lornaarmstrong — APPROVED** [review]
LGTM!

File: src/main/java/com/google/sps/data/User.java [hunk 5]
```diff
@@ -17,17 +17,21 @@
 /** The class represents user details. */
 public final class User {
 
-  private final String userEmail;
+  private final String email;
   private final String name;
   private final String department;
+  private final int year;
+  private final long phone;
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/data/User.java]: Would it be better to have it as a string? To allow leading zeros for example

File: src/main/java/com/google/sps/servlets/PersonalDataServlet.java [hunk 6]
```diff
@@ -40,20 +40,22 @@
...
+    String email = UserServiceFactory.getUserService().getCurrentUser().getEmail();
+    Key userKey = KeyFactory.createKey("User", email);
 
     try {
       // Get current user details from datastore
       Entity entity = DatastoreServiceFactory.getDatastoreService().get(userKey);
       String name = (String) entity.getProperty("name");
-      String department = (String) entity.getProperty("department");
+      String department = (String) entity.getProperty("department");      
+      int year = (int) (long) entity.getProperty("year");
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/servlets/PersonalDataServlet.java:53]: Why the double cast? `(int) entity.getProperty("year")` should be enough

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 7]
```diff
@@ -64,40 +64,47 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     // Convert entities to java class user entities
     List<User> users = new ArrayList<>();
     for (Entity entity : prepared.asIterable()) {
-      String userEmail = (String) entity.getProperty("userEmail");
+      String email = (String) entity.getProperty("email");
       String name = (String) entity.getProperty("name");
       String department = (String) entity.getProperty("department");
+      int year = (int) (long) entity.getProperty("year");
```
Comment (MostafaGabriel) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java:72]: ditto

File: src/main/webapp/form.html [hunk 8]
```diff
@@ -20,8 +20,17 @@ <h3>a recommendation system for students, by students!</h3>
...
       <input type="text" name="department" class="department" required></input>
+      <label for="year">currently in year</label>
+      <select name="year">
+        <option value="1">1</option>
+        <option value="2">2</option>
+        <option value="3">3</option>
+        <option value="4">4</option>
+      </select>
+      <label for="phone">my phone number is</label>
+      <input type="tel" id="phone" name="phone" required></input>
```
Comment (MostafaGabriel) [inline src/main/webapp/form.html]: Shall we use [the pattern parameter](https://www.w3schools.com/tags/att_input_type_tel.asp) here to standardize the format?

File: src/main/webapp/my-profile.html [hunk 9]
```diff
@@ -55,13 +55,22 @@ <h2>Edit Profile</h2>
...
               <input type="text" id="department" name="department" size="40" required></input>
+              <label for="year">currently in year</label>
+              <select id="year" name="year">
+                <option value="1">1</option>
+                <option value="2">2</option>
+                <option value="3">3</option>
+                <option value="4">4</option>
+              </select>
+              <label for="phone">my phone number is</label>
+              <input type="tel" id="phone" name="phone" required></input>
```
Comment (MostafaGabriel) [inline src/main/webapp/my-profile.html]: ditto

File: src/main/webapp/my-profile.html [hunk 10]
```diff
@@ -55,13 +55,22 @@ <h2>Edit Profile</h2>
               <input type="text" id="name" name="name" size="40" required></input>
               <label for="Department">I am currently a student in the department of</label>
```
Comment (MostafaGabriel) [inline src/main/webapp/my-profile.html]:
```suggestion
              <label for="Department">I am a student in the department of</label>
```

Just to match the one in the form

File: src/main/webapp/scripts/user-directory.js [hunk 11]
```diff
@@ -20,11 +20,12 @@ function generateUserCards() {
       userCardElement.className = "user-card";
       const profilePictureHTML = "<img class=\"user-img\" src=\"" + user.profilePictureUrl + "\"></img>";
```
Comment (MostafaGabriel) [inline src/main/webapp/scripts/user-directory.js]: nitty nit: can you please rename the css class to `profile-picture` to match HTML element name and also be consistent with the one in `my-profile`

**Review by MostafaGabriel — COMMENTED** [review]
Thanks Rania!

File: src/main/java/com/google/sps/servlets/PersonalDataServlet.java (hunk 6 above)
Comment (raniarusly) [reply src/main/java/com/google/sps/servlets/PersonalDataServlet.java:51]: apparently entity.getProperty returns an object and Casting it to int will gives an error but not to long.

File: src/main/java/com/google/sps/data/User.java (hunk 5 above)
Comment (raniarusly) [reply src/main/java/com/google/sps/data/User.java]: great idea, thank you for the suggestion :D

File: src/main/webapp/form.html (hunk 8 above)
Comment (raniarusly) [reply src/main/webapp/form.html]: This is important, especially when we are getting the telephone number as a string. I tried it but I am afraid it might be too strict, e.g. in the example referred we have to add hyphen and cannot input mobile number. I searched it up and there is no general convention for phone number as the convention differs by the [area](https://en.wikipedia.org/wiki/Telephone_numbers_in_the_United_Kingdom) or [country.](https://en.wikipedia.org/wiki/National_conventions_for_writing_telephone_numbers) What we can do is, we can restrict it to UK mobile phone number which has 11 digits, without restricting the spacing. Thus, the pattern will simply be pattern="[0-9]{11}"

File: src/main/java/com/google/sps/servlets/PersonalDataServlet.java (hunk 6 above)
Comment (MostafaGabriel) [reply src/main/java/com/google/sps/servlets/PersonalDataServlet.java:53]: Looks like Integer is [a supported data type](https://cloud.google.com/datastore/docs/concepts/entities#integer) by datastore. Can you please run `mvn package clean` to delete local datastore entities and see if the error still occurs? (I'm just worried that a `year` is stored there as a `long` that's why casting it as integer thrown an exception.

File: src/main/webapp/form.html (hunk 8 above)
Comment (MostafaGabriel) [reply src/main/webapp/form.html]: sounds good then! Thanks for looking into it :)

File: src/main/java/com/google/sps/servlets/PersonalDataServlet.java (hunk 6 above)
Comment (raniarusly) [reply src/main/java/com/google/sps/servlets/PersonalDataServlet.java:53]: Hi Mostafa, thank you for yor comment, so apparently entity.getProperty("department") returns an object
When casted directly to int, it gives an error
java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
However it can be casted to long from Long Object, and once its a long, it can be converted to int.
There are a few ways we can implement to int, one of the easiest is the double cast.
another one is use .intValue() but we have to case the object to Long first to call the fucntion, i.e. ((Long) entity.getProperty("year")).intValue()

**Review by MostafaGabriel — APPROVED** [review]
LGTM! Thanks Rania :)
